### DISCUSSION TITLE
Screen Magnification/Zooming

### DISCUSSION THREAD

**Discussion by miku4k** [YaLTeR/niri#601] [Ideas]
I just migrated from Hyprland and something I'm missing is a way to easily zoom into the screen. I rely on such functionality to comfortably view smaller details without having to lean too close to my screen (I'm a fan of maintaining my non-scoliosis status).

I thought a second about implementation details:
The zoom level should be controlled by an action, so there's an easy way to use it with keybinds, I'm thinking something similar to set-column-width functionality wise. 
If a configuration option exists, it should only control the zoom level on compositor startup or config reload. 

Also gotta watch out for naming conflicts with #352

**Comment by YaLTeR** [comment]
How should it work, zoom around the mouse cursor? Or have a dedicated area on the screen which shows zoomed-in contents?

**Reply by miku4k** [reply]
zooming in on the entire screen is what ive been thinking of, forgot to mention that.

**Reply by YaLTeR** [reply]
So around the mouse cursor?

**Reply by aquifolly** [reply]
i can think of multiple ways

1. small(ish) shape around the mouse cursor where what's in the shape is magnified (the cursor stays the same scale)
1. a static(?) box in a corner of the screen and that does a similar thing to the previous one (kind of like a camera?)
1. a full screen zoom in that follows the mouse

i believe these work in a similar way to the different modes of the windows magnifier

**Reply by miku4k** [reply]
I mean 3. It's what I'm used to and it's as far as I know also what most people use.

**Reply by rytswd** [reply]
Although this may not be exactly the same, I used to rely on a similar feature on macOS -- I found this post for Ubuntu / GNOME:
https://wiringbits.net/accessibility/2022/12/05/macos-like-pinch-to-zoom-on-ubuntu-22-04.html

This would be something I was hoping to get, which could be triggered with some key binding, or even touchpad pinching.

**Reply by bugi** [reply]
I asked on Matrix a few moments ago and was pointed to [wooz](https://github.com/negrel/wooz).

**Reply by zgibberish** [reply]
Ive used a ton and i like Plasma's and Cosmic's implementations the best, wooz always felt clunky to me

**Reply by bugi** [reply]
Thanks for the warning on wooz, so I won't get my hopes up too far.  I'm using Plasma currently specifically for occasional use of the zoom feature.  My eyes betray me!

I'd love to get back to the freedom of space I felt under i3, and so far niri's strip feels pretty much how I used i3 but on purpose.

**Reply by miku4k** [reply]
the problem for me with all zooming tools is that they work with screenshots (ie nothing moves when youre zoomed in), which doesnt fit my use case

**Comment by knghtbrd** [comment]
I'm interested in this and, actually I got real excited about using niri on my 2-in-1 tablet until I discovered this basically doesn't exist currently. The usual modern ways I see zoom implemented are these:

- Zoom continuously with pointer
- Cursor pushes viewport around the screen
- Magnifier lens moves with cursor
- Portion of the screen (bar typically) is magnified

Of these IMO the second always feels laggy, and the third and fourth are dumb (especially since the last one usually treats the magnifier as a window on the screen and does the hall of mirrors thing. The first one works the best.

It's important to be able to control the zoom pretty finely because you need to use the least zoom possible or find using the whole machine frustrating. And actually, yes, there are other options, but sometimes a momentary zoom is exactly the right solution. What I mean is … my desktop has a 38" 3840x1600 monitor using 192 DPI under XFCE. So not a proper HiDPI scale, just fonts. But that's _usually_ what I want, actually. In Firefox, I have Github scaled to 120% size and because I'm tired I've zoomed in on this edit box with fullscreen zoom to about 1.4x, which is high for me.

Why not just HiDPI the screen? Well, I was talking about my tablet. So, Mobian … and phosh … with 2x HiDPI on a 12" 1080p screen. *takes a breath* If I have any window open and pop the menu, the vertical stacking goeth thusly:

1. The top bar, over 1cm tall
2. The firefox window I have open, takes up half the screen
3. The search bar, 30% taller than it needs to be, but can't change it
4. About 2.5 rows of icons, these are what I actually want right now, scrollscrollscrollscrollscroll
5. A filter between touch-optimized and all apps, why can't this be with the search bar?

If this were an iPad or Android, the top bar would be normal sized (so, too small to be convenient at a "normal" distance, but I could expressly move my head down to the screen or use screen magnification to see it), the screen thumbnails and icons would be divorced from the font size, icon names would probably be cut off and useless or left honestly much too small depending on how I adjusted the two sliders that control font sizes in the OS, I'd have a lot more icons on the home screen at once (and identifying programs by icon rather than name), and there'd be a widget on the screen showing time, next alarm, and battery percentage. In other words, no magnification vs. normal except for screen zoom and using 2x2 of icon space for the crap the mostly useless top bar should be telling me. (Ironically I magnify my phone the least because I can hold it up to my nose, but I can't actually use it as well as others can for most things.

On any other project I'd file a ticket about this, but … I learned my lesson with Gnome literally 11 years ago, and every few days or so I read a new report that not much has changed. I was really only examining the state of the art for touch devices to determine if I needed to roll my own solution or if Plasma would suffice (or if phosh was really that far ahead of Plasma… Ehh, I probably still do want my own solution if only because the at least 528 or so Debian Qt packages are just a little bit ridiculous. (No, that's not actually an exaggeration, I checked a Debian box for trixie…)

**Comment by Schweeeeeeeeeeeeeeee** [comment]
I really love the kde zoom just press meta plus and it zooms in on my cursor would really love the feature

**Comment by knghtbrd** [comment]
The first time I used a full-screen variable zoom it was on MacOS X 10.3 onward and it was honestly the killer app for a guy with limited vision. I bought a Mac because of that feature and I used one as my primary workstation for a long time. By the time I came back to the Linux desktop, Xorg had fully replaced XFree86, the old way of switching to a low-ass resolution slowly to magnify the screen on a virtual desktop didn't seem to even be an option anymore, and there were these compositing WMs that could do the same thing the Mac did. Just Cmd + Option + =/- to zoom iin or out, or configurable modifier + scroll wheel to zoom the entire screen around the cursor in fine steps. The current state of this on Linux AFAIK:

- Gnome has it with mod + wheel. Technically with same keys as Mac too (Super + Alt + =/-) but I haven't checked back in 11 years since I was told I'm not supposed to use my computer that way. I'll never get an apology for that, so I'll never forgive nor forget.

- Cinnamon is like Gnome, except keyboard and mouse both work in small/smooth steps.

- KDE has both keyboard and mouse. Pretty extensively configurable. I didn't use this on Debian sid because Debian Qt/KDE package counts/interdependencies are 🤬ing terrifying. It MAY have the option to control zoom using screen or touchpad gestures ala iOS too, actually, I've read that someone set this up. Of course I can't figure out how to do it on a mobian plasma 13 pre-release but KDE seems to have at least two overlapping settings applications to really give you that Windows experience. 😁 I can report back on this if anyone cares.

- XFCE has mod + mouse only, and the mod is only half-configurable. Whatever you set to be the key you use to e.g. drag windows around sans titlebar will also be the mod for zoom. I set that to Super and then bind Super + =/- to send fake wheel events. This works almost perfectly. Some apps ignore these fake wheel events because they're bound to do something, but others do not. Firefox is in the do not category, which is annoying.

- MATE does not have this feature at all because marco is really is just a sad Gnome 2 Metacity fork with almost no development and certainly no compositor. It's common to sub Marco out for Compiz, but if you know how to use dconf-editor or gsettings you can swap it for anything. I did this myself, and then discovered that Linux Mint MATE Edition does this by default, for this reason. Good on ya, Mint MATEs!

- LXDE doesn't support this, but again xfwm4 is possible.  And LXQt can use Kwin.

- Compiz definitely does this, but it's a buggy, crashy mess and has been since before it was Compiz.

- Weston does it(!) Yes, really!

- Wayfire had at least mod + wheel support this time last year. No keyboard control at the time. It wouldn't be hard to add, but nobody had at the time. No idea if that's changed.

- labwc has zoom support you can bind to whatever you like. You have to write the XML to make it do any of that yourself, of course, but XML but if you're here I assume you can handle that task.

- Hyprland has support. To what extend I don't know because I'm sitting here looking at a 38" screen with a maximized Firefox window. One of I'm not going to count that many windows. I don't think a traditional tiling compositor is for me, at least not on desktop. I'd be fine with niri's "metroidvania" scrolling since I could e.g. collapse this pile of fullscreen firefox windows into one set of "tabs" and call it a day.

Basically this is a widely adopted accessibility feature, but it's certainly not universal, and there's a couple of implementations for how to do it on the keyboard, but mod + wheel is almost ubiquitous. It's usually configurable at least somewhat which mod is used, and when it is directly configured in isolation ctrl + wheel seems to be the default. Keyboard zoom control either uses super + alt or just super with = and - to zoom in and out.

Separately, but hand in hand with the zoom feature, often is a couple of pixel shader filters: Grayscale everything and color invert (a simple one minus source color, which is why the default for turning it on also turns on the grayscale feature. The two filters can be decoupled though even on modern macOS.) It's not very widely implemented on Linux, but with dark themes and darkreader, it's less important IMO. Of course dark mode is only really necessary for most people because some idiot decided it was trendy to flashbang users on a regular basis with thin black fonts on face-melting white voids as a (non)"sensible" default. Neutral gray midtone windows are so 1990s, and soothing colors and graphics effects are so ten years ago. Now we're going to have **HDR** so you can tell how far away the flashbang is—point blank, it's always gonna be point blank!

So yeah, that's the state of things, basically.

**Reply by YaLTeR** [reply]
Thanks, this is quite a comprehensive overview

**Comment by ASong5** [comment]
KWin offers this feature natively and was what I used quite often when I was on KDE. I hope this is something that can be added in the near future.

**Comment by ricostap** [comment]
Just to put my vote down, this is probably the only feature I am missing when comparing Niri with Hyprland.

**Comment by olejorgenb** [comment]
Another downside of external implementations like `wooz` is that they don't support (can't?) zooming _out_. Not as commonly useful, but I think zooming out can sometimes be useful.

**Comment by nickjj** [comment]
For folks who want this feature right now for full screen zoom, unless I'm misunderstanding something we can do it with an output's scale right?

For example you can run `msg output DP-1 scale 2.0`, replacing `DP-1` with your output's name and now you're zoomed in 2x.

From there you could wire up a script that inspects the output's current scale and then +/- it based on the direction you want to zoom (in or out) and lastly you could bind this to invoke it through hotkeys. In the end this lets you incrementally zoom in and out by 0.1 or whatever increment you want to define.

Or are you hoping this gets built into niri so you don't have to wire up a custom solution?

**Reply by JacksonChen666** [reply]
It is not display scaling, that's essentially fit to scale. It's zooming, like taking a screenshot and zooming it in an image editor, but with much less friction of taking a screenshot and opening it in an image viewer, and also doing it live.

**Comment by B83C** [comment]
Something like cosmic desktop's zooming. That one was pretty decent. Win + Mouse scroll to zoom in/out.
